### ISSUE TITLE
Telegraf - Input Modbus Plugin crashes Telegraf if it hits an "illegal address"

### ISSUE THREAD

**Issue by spkis** [influxdata/telegraf#14671]
### Relevant telegraf.conf

```toml
[global_tags]

[agent]
  interval = "10s"
  round_interval = true
  metric_batch_size = 1000
  metric_buffer_limit = 10000
  collection_jitter = "0s"
  flush_interval = "10s"
  precision = "1ms"
  #debug = true
  # quiet = false
  # logtarget = "file"
  #logfile = "/var/log/telegraf/debug.log"
  hostname = ""
  omit_hostname = true

[[outputs.file]]
  files = ["/tmp/telegraf/metrics.out"]
  data_format = "influx"

# [[outputs.influxdb_v2]]
#   urls = ["http://XXXXX:9086"]
#   token = "TOKLI"
#   organization = "COMPANY"
#   bucket = "samsung_batteries"
#   # bucket_tag = ""
#   # exclude_bucket_tag = false
#   # timeout = "5s"
#   # http_headers = {"X-Special-Header" = "Special-Value"}
#   # http_proxy = "http://corporate.proxy:3128"
#   # user_agent = "telegraf"
#   # content_encoding = "gzip"
#   ## Enable or disable uint support for writing uints influxdb 2.0.
#   influx_uint_support = true

#   ## Optional TLS Config for use on HTTP connections.
#   # tls_ca = "/etc/telegraf/ca.pem"
#   # tls_cert = "/etc/telegraf/cert.pem"
#   # tls_key = "/etc/telegraf/key.pem"
#   ## Use TLS but skip chain & host verification
#   #insecure_skip_verify = false

[[inputs.modbus]]
  name = "prod"
  timeout = "3s"
  # TCP - connect via Modbus/TCP
  #controller = "tcp://xxxx:602"
  controller = "tcp://xxxxx:602"
  #debug_connection = true
  configuration_type = "request"


#Rack1
[[inputs.modbus.request]]
    slave_id = 1
   # optimization = "max_insert"
   # optimization_max_register_fill = 50
    byte_order = "ABCD"
    register = "input"
    fields = [
# Rack Level
{ measurement = "Battery", name = "Rack Voltage", type = "UINT16", scale=0.1, address = 10200 },
{ measurement = "battery", name = "Mode", type = "UINT16", scale = 1.0, address = 1360 },
        { measurement = "battery", name = "Alarm #1", type = "UINT16", scale = 1.0, address = 1362 },
        { measurement = "battery", name = "Alarm #2", type = "UINT16", scale = 1.0, address = 1363 },
        { measurement = "battery", name = "Alarm #3", type = "UINT16", scale = 1.0, address = 1364 },
        { measurement = "battery", name = "Alarm #4", type = "UINT16", scale = 1.0, address = 1365 },
        { measurement = "battery", name = "Protection #1", type = "UINT16", scale = 1.0, address = 1366 },
        { measurement = "battery", name = "Protection #2", type = "UINT16", scale = 1.0, address = 1367 },
        { measurement = "battery", name = "Protection #3", type = "UINT16", scale = 1.0, address = 1368 },
        { measurement = "battery", name = "Protection #4", type = "UINT16", scale = 1.0, address = 1369 },
        { measurement = "battery", name = "Control Status", type = "UINT16", scale = 1.0, address = 1370 },
        { measurement = "battery", name = "Voltage (String Sensing)", type = "UINT16", scale = 0.1, address = 1373 },
        { measurement = "battery", name = "Voltage (Module Sensing Summation)", type = "UINT16", scale = 0.1, address = 1374 },
        { measurement = "battery", name = "Voltage (Cell Sensing Summation)", type = "UINT16", scale = 0.1, address = 1375 },
        { measurement = "battery", name = "Current (Sensing)", type = "INT16", scale = 0.1, address = 1376 },
        { measurement = "battery", name = "Current (Average for 3 seconds)", type = "INT16", scale = 0.1, address = 1377 },
        { measurement = "battery", name = "SOC (State of Charge)", type = "UINT16", scale = 0.1, address = 1378 },
        { measurement = "battery", name = "SOH (State of Health)", type = "UINT16", scale = 0.1, address = 1379 },
        { measurement = "battery", name = "Current Limit of Charge", type = "UINT16", scale = 0.1, address = 1380 },
        { measurement = "battery", name = "Current Limit of Discharge", type = "UINT16", scale = 0.1, address = 1381 },
        { measurement = "battery", name = "Power Limit of Charge", type = "UINT16", scale = 10.0, address = 1382 },
        { measurement = "battery", name = "Power Limit of Discharge", type = "UINT16", scale = 10.0, address = 1383 },
        { measurement = "battery", name = "Accumulated Amount of Charge Power", type = "UINT16", scale = 1.0, address = 1384 },
        { measurement = "battery", name = "Accumulated Amount of Discharge Power", type = "UINT16", scale = 1.0, address = 1388 },
        { measurement = "battery", name = "Module Power Off Position", type = "UINT16", scale = 1.0, address = 1396 },
        { measurement = "battery", name = "Module Communication Fault Position", type = "UINT16", scale = 1.0, address = 1397 },
        { measurement = "battery", name = "Cell Voltage (Max Difference)", type = "UINT16", scale = 0.001, address = 1398 },
        { measurement = "battery", name = "Cell Voltage (Average)", type = "UINT16", scale = 0.001, address = 1399 },
        { measurement = "battery", name = "Cell Voltage (Max(1st))", type = "UINT16", scale = 0.001, address = 1400 },
        { measurement = "battery", name = "Cell Voltage (Max(1st)) Module Position", type = "UINT16", scale = 1.0, address = 1401 },
        { measurement = "battery", name = "Cell Voltage (Max(1st)) Cell Position", type = "UINT16", scale = 1.0, address = 1402 },
        { measurement = "battery", name = "Cell Voltage (Min(1st))", type = "UINT16", scale = 0.001, address = 1403 },
        { measurement = "battery", name = "Cell Voltage (Min(1st)) Module Position", type = "UINT16", scale = 1.0, address = 1404 },
        { measurement = "battery", name = "Cell Voltage (Min(1st)) Cell Position", type = "UINT16", scale = 1.0, address = 1405 },
        { measurement = "battery", name = "Cell Voltage (Max(2nd))", type = "UINT16", scale = 0.001, address = 1406 },
        { measurement = "battery", name = "Cell Voltage (Max(2nd)) Module Position", type = "UINT16", scale = 1.0, address = 1407 },
        { measurement = "battery", name = "Cell Voltage (Max(2nd)) Cell Position", type = "UINT16", scale = 1.0, address = 1408 },
        { measurement = "battery", name = "Cell Voltage (Min(2nd))", type = "UINT16", scale = 0.001, address = 1409 },
        { measurement = "battery", name = "Cell Voltage (Min(2nd)) Module Position", type = "UINT16", scale = 1.0, address = 1410 },
        { measurement = "battery", name = "Cell Voltage (Min(2nd)) Cell Position", type = "UINT16", scale = 1.0, address = 1411 },
        { measurement = "battery", name = "Cell Temperature (Max Difference)", type = "UINT16", scale = 0.01, address = 1412 },
        { measurement = "battery", name = "Cell Temperature (Average)", type = "INT16", scale = 0.01, address = 1413 },
        { measurement = "battery", name = "Cell Temperature (Max(1st))", type = "INT16", scale = 0.01, address = 1414 },
        { measurement = "battery", name = "Cell Temperature (Max(1st)) Module Position", type = "UINT16", scale = 1.0, address = 1415 },
        { measurement = "battery", name = "Cell Temperature (Max(1st)) Cell Position", type = "UINT16", scale = 1.0, address = 1416 },
        { measurement = "battery", name = "Cell Temperature (Min(1st))", type = "INT16", scale = 0.01, address = 1417 },
        { measurement = "battery", name = "Cell Temperature (Min(1st)) Module Position", type = "UINT16", scale = 1.0, address = 1418 },
        { measurement = "battery", name = "Cell Temperature (Min(1st)) Cell Position", type = "UINT16", scale = 1.0, address = 1419 },
        { measurement = "battery", name = "Cell Temperature (Max(2nd))", type = "INT16", scale = 0.01, address = 1420 },
        { measurement = "battery", name = "Cell Temperature (Max(2nd)) Module Position", type = "UINT16", scale = 1.0, address = 1421 },
        { measurement = "battery", name = "Cell Temperature (Max(2nd)) Cell Position", type = "UINT16", scale = 1.0, address = 1422 },
        { measurement = "battery", name = "Cell Temperature (Min(2nd))", type = "INT16", scale = 0.01, address = 1423 },
        { measurement = "battery", name = "Cell Temperature (Min(2nd)) Module Position", type = "UINT16", scale = 1.0, address = 1424 },
        { measurement = "battery", name = "Cell Temperature (Min(2nd)) Cell Position", type = "UINT16", scale = 1.0, address = 1425 },
        { measurement = "battery", name = "Module PCB Temperature (Max)", type = "INT16", scale = 0.01, address = 1426 },
        { measurement = "battery", name = "Module PCB Temperature (Max) Module Position", type = "UINT16", scale = 1.0, address = 1427 },
        { measurement = "battery", name = "Module PCB Temperature (Min)", type = "INT16", scale = 0.01, address = 1428 },
        { measurement = "battery", name = "Module PCB Temperature (Min) Module Position", type = "UINT16", scale = 1.0, address = 1429 },
        { measurement = "battery", name = "Work Cell Index", type = "UINT16", scale = 1.0, address = 1434 },
        { measurement = "battery", name = "Sensor Status Register", type = "UINT16", scale = 1.0, address = 1435 },
        { measurement = "battery", name = "Module Status Register (Scan)", type = "UINT16", scale = 1.0, address = 1436 },
        { measurement = "battery", name = "Module Status Register (Balancing)", type = "UINT16", scale = 1.0, address = 1437 },
        { measurement = "battery", name = "Module Status Register (Address)", type = "UINT16", scale = 1.0, address = 1438 },
        { measurement = "battery", name = "Module Status Register (Voltage)", type = "UINT16", scale = 1.0, address = 1439 },
        { measurement = "battery", name = "Module Status Register (Temperature)", type = "UINT16", scale = 1.0, address = 1440 },
        { measurement = "battery", name = "Difference Voltage (Max(Cell Sensing Summation))", type = "UINT16", scale = 0.01, address = 1441 },
        { measurement = "battery", name = "Difference Voltage (Max(Module Sensing))", type = "UINT16", scale = 0.01, address = 1442 },
        { measurement = "battery", name = "Difference Voltage (Max) Module Position", type = "UINT16", scale = 1.0, address = 1443 },
        { measurement = "battery", name = "Difference Voltage (Min(Cell Sensing Summation))", type = "UINT16", scale = 0.01, address = 1444 },
        { measurement = "battery", name = "Difference Voltage (Min(Module Sensing))", type = "UINT16", scale = 0.01, address = 1445 },
        { measurement = "battery", name = "Difference Voltage (Min) Module Position", type = "UINT16", scale = 1.0, address = 1446 },
        { measurement = "battery", name = "State", type = "UINT16", scale = 1.0, address = 1447 },
        { measurement = "battery", name = "Cell V Distribution (Under 20mV)", type = "UINT16", scale = 1.0, address = 1448 },
        { measurement = "battery", name = "Cell V Distribution (Under V-IMB Alarm Level)", type = "UINT16", scale = 1.0, address = 1449 },
        { measurement = "battery", name = "Cell V Distribution (Under V-IMB Protection Level)", type = "UINT16", scale = 1.0, address = 1450 },
        { measurement = "battery", name = "Cell V Distribution (V-IMB Protection Level or More)", type = "UINT16", scale = 1.0, address = 1451 },
        { measurement = "battery", name = "Cell Balancing Execution Count (Max)", type = "UINT16", scale = 1.0, address = 1452 },
        { measurement = "battery", name = "Cell Balancing Execution Count (Max) Module Position", type = "UINT16", scale = 1.0, address = 1453 },
        { measurement = "battery", name = "Cell Balancing Execution Count (Max) Cell Position", type = "UINT16", scale = 1.0, address = 1454 },
        { measurement = "battery", name = "Max Balancing Time (minute)", type = "UINT16", scale = 1.0, address = 1455 },
        { measurement = "battery", name = "Current (Sensing_Sub)", type = "INT16", scale = 0.1, address = 1457 },
        { measurement = "battery", name = "Min Cell Deviation Index", type = "INT16", scale = 1.0, address = 1458 },
        { measurement = "battery", name = "Min Cell Deviation Index Module Position", type = "UINT16", scale = 1.0, address = 1459 },
        { measurement = "battery", name = "Min Cell Deviation Index Cell Position", type = "UINT16", scale = 1.0, address = 1460 },
        { measurement = "battery", name = "SOC (State of Charge(Ref))", type = "INT16", scale = 0.1, address = 1461 },
        { measurement = "battery", name = "Module Ambient Temperature (Max)", type = "INT16", scale = 0.01, address = 1466 },
        { measurement = "battery", name = "Module Ambient Temperature (Max) Module Position", type = "UINT16", scale = 1.0, address = 1467 },
        { measurement = "battery", name = "Module Ambient Temperature (Max) Sensor Position", type = "UINT16", scale = 1.0, address = 1468 },
        { measurement = "battery", name = "Module Ambient Temperature (Min)", type = "INT16", scale = 0.01, address = 1469 },
        { measurement = "battery", name = "Module Ambient Temperature (Min) Module Position", type = "UINT16", scale = 1.0, address = 1470 },
        { measurement = "battery", name = "Module Ambient Temperature (Min) Sensor Position", type = "UINT16", scale = 1.0, address = 1471 },
        { measurement = "battery", name = "Module PCB Temperature (Max) Sensor Position", type = "UINT16", scale = 1.0, address = 1472 },
        { measurement = "battery", name = "Module PCB Temperature (Min) Sensor Position", type = "UINT16", scale = 1.0, address = 1473 },

]
```

### Logs from Telegraf

```text
2024-02-02T10:59:51Z D! [inputs.modbus] trying to read input@1373[12]...
2024-02-02T10:59:51Z D! [inputs.modbus] got input@1373[12]: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 100 0 5 0 48]
2024-02-02T10:59:51Z D! [inputs.modbus]   field Voltage (String Sensing) with offset 0 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Voltage (Module Sensing Summation) with offset 2 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Voltage (Cell Sensing Summation) with offset 4 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Current (Sensing) with offset 6 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Current (Average for 3 seconds) with offset 8 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field SOC (State of Charge) with offset 10 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field SOH (State of Health) with offset 12 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Current Limit of Charge with offset 14 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Current Limit of Discharge with offset 16 with len 2: [0 0] --> 0
2024-02-02T10:59:51Z D! [inputs.modbus]   field Power Limit of Charge with offset 18 with len 2: [0 100] --> 1000
2024-02-02T10:59:51Z D! [inputs.modbus]   field Power Limit of Discharge with offset 20 with len 2: [0 5] --> 50
2024-02-02T10:59:51Z D! [inputs.modbus]   field Accumulated Amount of Charge Power with offset 22 with len 2: [0 48] --> 48
2024-02-02T10:59:51Z D! [inputs.modbus] trying to read input@1388[1]...
2024-02-02T10:59:51Z D! [inputs.modbus] got input@1388[1]: [48 48]
2024-02-02T10:59:51Z D! [inputs.modbus]   field Accumulated Amount of Discharge Power with offset 0 with len 2: [48 48] --> 12336
2024-02-02T10:59:51Z D! [inputs.modbus] trying to read input@1396[34]...
2024-02-02T10:59:51Z E! [inputs.modbus] Error in plugin: slave 1: modbus: exception '2' (illegal data address), function '132'
2024-02-02T10:59:51Z D! [inputs.modbus] Reconnecting to tcp://10.0.32.11:602...
2024-02-02T10:59:51Z D! [agent] Stopping service inputs
2024-02-02T10:59:51Z D! [agent] Input channel closed
2024-02-02T10:59:51Z D! [agent] Stopped Successfully
2024-02-02T10:59:51Z E! [telegraf] Error running agent: input plugins recorded 1 errors
```

### System info

Telegraf 1.29.4, Linux Native, Mac Native, Docker, doesn't make a difference :) 

### Steps to reproduce

1. Generate your telegraf.conf config. 
2. Try to debug and check your work.
3. Notice you run into an exception, "illegal address"
4. Do the same registers with simple python code, the code does NOT run in an expeption and stops. 
Here is an example script:
`from pyModbusTCP.client import ModbusClient

 

client = ModbusClient('xxxx', 602, auto_open=True)

client.debug = True

register = [1395]

 

for i in register:

    print(client.read_input_registers(i))

 

 

print(client.read_input_registers(1395))  `

### Expected behavior

Telegraf does not crash.

I would want it still runs and continue on. 
It would be cool if Telegraf could add information (like a tag or field) with affected (non working) fields so that you can easily see what addresses are failing.

Also it would be great if you could maybe set per configuration, what data should go in the field which is not able to gather information from.

### Actual behavior

It crashes and I don't get ANY data - not even from the hundreds of registers which work :) 

In reality, it could be that 1 from 11 racks is defective and needs to be turned off.
I would get the error now - and don't get any data from 10 racks anymore until I change the configuration to work around the currently defective 1 rack.

Hence I think this should be handled differently/better :-)

**Comment by srebhan** [comment]
@spkis your python code reads register `1395` but your Telegraf config specifies `1396`!
> 2024-02-02T10:59:51Z D! [inputs.modbus] trying to read input@1396[34]...
> 2024-02-02T10:59:51Z E! [inputs.modbus] Error in plugin: slave 1: modbus: exception '2' (illegal data address), function '132'

**Comment by spkis** [comment]
@srebhan 
Sorry for that - using the same register I get in python also an error - but the code just doesn't crash:

Using this sample registers:

`register = [1395,1396,2433]

 

for i in register:

    print(client.read_input_registers(i))

 `

outputs to
`Tx
[BD F6 00 00 00 06 01] 04 05 73 00 01
Rx
[BD F6 00 00 00 03 01] 84 02
modbus exception (code 2 "illegal data address")
None
Tx
[CB B4 00 00 00 06 01] 04 05 74 00 01
Rx
[CB B4 00 00 00 03 01] 84 02
modbus exception (code 2 "illegal data address")
None
Tx
[C5 24 00 00 00 06 01] 04 09 81 00 01
Rx
[C5 24 00 00 00 05 01] 04 02 00 00
[0]`

Telegraf however stops at register 1396 - and doesn't even try to reach register 2433.

**Comment by srebhan** [comment]
@spkis Telegraf itself does not stop, it will only stop the gathering for **this one plugin**! That's a big difference!
Illegal data addresses are a misconfiguration and therefore we follow the "fail-early-and-loudly" philosophy. This way you _notice_ that there is an issue where otherwise you would assume everything is fine and then wonder after some month why your one field is missing!

Furthermore, Telegraf optimizes the requests and tries to read 1396 **plug the next 34 registers** in one request. This is far more efficient compared to what you do in your python code! How should we know which register to skip in this bunch of 34?

**Comment by spkis** [comment]
My problem is, that in reality you might end up losing a rack or a rack is not (yet) updated.
With this current workflow it will just stop to collect data if it hits an error.
The python code just continues on.

With telegraf modbus plugin failing, I then get no data at all which is a problem for us. 
=> The python is not optimized, however it will just not stop collecting data. Telegraf is.

**Comment by spkis** [comment]
Hi, 
I forked telegraf and added an addtional flag into the configuration. I made a push request for this. 

Since other plugins also have skip errors, I think from a design perspective it would be making sense to allign this adapter with this option as well.
Additionally, I don't want to move away from Telegraf - while I know that it's not per se a telegraf issue :-)

**Comment by spkis** [comment]
From discussion in the Pull Request - hence closing this for now:

"I did research internally. Apparently, we get the batteries already delivered with different firmware images on the racks.
I was able to reverse engineer the configuration files of the devices and luckily, we have some new modbus registers which are relevant for warranty in newer firmware versions.
Hence we are trying to get the firmware updates again with more pressure now.
=> The different versions are not from production updates, hence the issue I tried to workaround, isn't needed and I will try to workaround those racks upfront."
